File: futureagi/tracer/services/clickhouse/v2/property_catalog/producer_retirement.py
"""Durable activation proof for bounded Go hot-producer state retirement."""

from __future__ import annotations

import json
import os
import stat
import tempfile
from collections.abc import Mapping, Sequence
from dataclasses import dataclass, replace
from datetime import UTC, datetime
from pathlib import Path
from typing import Any

from .activation import CatalogLifecycleMode, ManifestStreamRole, RevisionBuildPlan
from .codec import canonical_uuid, framed_sha256, require_sha256
from .durable_lifecycle import PriorActiveEvidence
from .models import SourceAdapter
from .runtime_limits import RUNTIME_LIMITS

PRODUCER_RETIREMENT_FORMAT = "futureagi.property-catalog-producer-state-retirements"
PRODUCER_RETIREMENT_VERSION = 1
PRODUCER_RETIREMENT_FILE_NAME = "producer-state-retirements-v1.json"
PRODUCER_RETIREMENT_SHA_DOMAIN = (
    "futureagi.property-catalog.producer-state-retirement.v1"
)
MAX_PRODUCER_RETIREMENTS = 256
MAX_PRODUCER_RETIREMENT_BYTES = RUNTIME_LIMITS.producer_retirement_max_bytes
_TIME_FORMAT = "%Y-%m-%d %H:%M:%S.%f"
_ZERO_SHA256 = "0" * 64
_RECORD_FIELDS = (
    "organization_id",
    "workspace_id",
    "catalog_epoch",
    "catalog_revision",
    "build_token",
    "projection_version",
    "lifecycle_mode",
    "build_plan_json",
    "build_lease_sha256",
    "source_manifest_sha256",
    "activation_sequence",
    "activation_sha256",
    "lineage_anchor_revision",
    "lineage_anchor_build_token",
    "lineage_anchor_activation_sequence",
    "lineage_anchor_activation_sha256",
    "active_revisions_since_anchor",
    "hot_producer_stream_id",
    "emitted_at",
    "retirement_sha256",
)


class ProducerRetirementError(RuntimeError):
    """Activation retirement evidence is missing, ambiguous, or unsafe."""


def _positive_uint(value: object, bits: int, field: str) -> int:
    if type(value) is not int or not 1 <= value < (1 << bits):
        raise ProducerRetirementError(f"{field} must be a positive UInt{bits}")
    return value


def _time_text(value: datetime) -> str:
    if not isinstance(value, datetime) or value.tzinfo is None:
        raise ProducerRetirementError("emitted_at must be timezone-aware")
    normalized = value.astimezone(UTC)
    if value.utcoffset() != UTC.utcoffset(value):
        raise ProducerRetirementError("emitted_at must use UTC")
    return normalized.strftime(_TIME_FORMAT)


def _parse_time(value: object) -> datetime:
    if not isinstance(value, str):
        raise ProducerRetirementError("emitted_at must be canonical text")
    try:
        parsed = datetime.strptime(value, _TIME_FORMAT).replace(tzinfo=UTC)
    except ValueError as exc:
        raise ProducerRetirementError("emitted_at must be canonical UTC") from exc
    if _time_text(parsed) != value:
        raise ProducerRetirementError("emitted_at must be canonical UTC")
    return parsed


@dataclass(frozen=True, slots=True)
class ProducerStateRetirement:
    organization_id: str
    workspace_id: str
    catalog_epoch: int
    catalog_revision: int
    build_token: str
    projection_version: int
    lifecycle_mode: CatalogLifecycleMode
    build_plan_json: str
    build_lease_sha256: str
    source_manifest_sha256: str
    activation_sequence: int
    activation_sha256: str
    lineage_anchor_revision: int
    lineage_anchor_build_token: str
    lineage_anchor_activation_sequence: int
    lineage_anchor_activation_sha256: str
    active_revisions_since_anchor: int
    hot_producer_stream_id: str
    emitted_at: datetime
    retirement_sha256: str = _ZERO_SHA256

    def __post_init__(self) -> None:
        for field in (
            "organization_id",
            "workspace_id",
            "build_token",
            "lineage_anchor_build_token",
            "hot_producer_stream_id",
        ):
            object.__setattr__(
                self,
                field,
                canonical_uuid(getattr(self, field), field=field),
            )
        _positive_uint(self.catalog_epoch, 16, "catalog_epoch")
        _positive_uint(self.catalog_revision, 64, "catalog_revision")
        _positive_uint(self.projection_version, 16, "projection_version")
        _positive_uint(self.activation_sequence, 64, "activation_sequence")
        _positive_uint(
            self.lineage_anchor_revision,
            64,
            "lineage_anchor_revision",
        )
        _positive_uint(
            self.lineage_anchor_activation_sequence,
            64,
            "lineage_anchor_activation_sequence",
        )
        if type(self.active_revisions_since_anchor) is not int or not (
            0 <= self.active_revisions_since_anchor <= 2_048
        ):
            raise ProducerRetirementError(
                "active_revisions_since_anchor exceeds its lifecycle bound"
            )
        if not isinstance(self.lifecycle_mode, CatalogLifecycleMode):
            raise ProducerRetirementError("lifecycle_mode is invalid")
        for field in (
            "build_lease_sha256",
            "source_manifest_sha256",
            "activation_sha256",
            "lineage_anchor_activation_sha256",
        ):
            require_sha256(getattr(self, field), field=field)
        _time_text(self.emitted_at)
        try:
            plan = RevisionBuildPlan.from_json(self.build_plan_json)
        except (TypeError, ValueError) as exc:
            raise ProducerRetirementError(
                "retirement build_plan_json is not canonical v2"
            ) from exc
        if (
            plan.organization_id != self.organization_id
            or plan.workspace_id != self.workspace_id
            or plan.catalog_epoch != self.catalog_epoch
            or plan.catalog_revision != self.catalog_revision
            or plan.build_token != self.build_token
            or plan.projection_version != self.projection_version
            or plan.sha256 != self.build_lease_sha256
        ):
            raise ProducerRetirementError(
                "retirement identity differs from its canonical build plan"
            )
        hot = tuple(
            stream
            for stream in plan.streams
            if stream.source_adapter is SourceAdapter.SPAN_ATTRIBUTE
            and stream.role is ManifestStreamRole.HOT_VALUES
        )
        if len(hot) != 1 or hot[0].producer_stream_id != self.hot_producer_stream_id:
            raise ProducerRetirementError(
                "retirement hot stream differs from the active build plan"
            )
        if (
            self.activation_sequence - self.lineage_anchor_activation_sequence
            != self.active_revisions_since_anchor
        ):
            raise ProducerRetirementError(
                "retirement activation depth differs from its lineage anchor"
            )
        snapshot_mode = self.lifecycle_mode in {
            CatalogLifecycleMode.INITIAL_BACKFILL,
            CatalogLifecycleMode.FULL_REPAIR,
        }
        if snapshot_mode and (
            self.lineage_anchor_revision != self.catalog_revision
            or self.lineage_anchor_build_token != self.build_token
            or self.lineage_anchor_activation_sequence != self.activation_sequence
            or self.lineage_anchor_activation_sha256 != self.activation_sha256
            or self.active_revisions_since_anchor != 0
        ):
            raise ProducerRetirementError(
                "snapshot retirement is not its own lineage anchor"
            )
        if not snapshot_mode and (
            self.lineage_anchor_revision >= self.catalog_revision
            or self.lineage_anchor_activation_sequence >= self.activation_sequence
        ):
            raise ProducerRetirementError(
                "incremental retirement has no earlier lineage anchor"
            )
        expected = producer_retirement_sha256(self)
        if self.retirement_sha256 == _ZERO_SHA256:
            object.__setattr__(self, "retirement_sha256", expected)
        elif self.retirement_sha256 != expected:
            raise ProducerRetirementError("retirement digest does not match its fields")

    @classmethod
    def from_active(
        cls,
        active: PriorActiveEvidence,
        *,
        organization_id: str,
        workspace_id: str,
        catalog_epoch: int,
        emitted_at: datetime,
    ) -> ProducerStateRetirement:
        if not isinstance(active, PriorActiveEvidence):
            raise TypeError("active must be PriorActiveEvidence")
        plan = active.build_plan
        if (
            plan.organization_id
            != canonical_uuid(
                organization_id,
                field="organization_id",
            )
            or plan.workspace_id != canonical_uuid(workspace_id, field="workspace_id")
            or plan.catalog_epoch != catalog_epoch
        ):
            raise ProducerRetirementError(
                "active evidence differs from the requested retirement scope"
            )
        hot = active.stream(
            SourceAdapter.SPAN_ATTRIBUTE,
            ManifestStreamRole.HOT_VALUES,
        )
        anchor = active.lineage_anchor
        return cls(
            organization_id=plan.organization_id,
            workspace_id=plan.workspace_id,
            catalog_epoch=plan.catalog_epoch,
            catalog_revision=active.catalog_revision,
            build_token=active.build_token,
            projection_version=active.projection_version,
            lifecycle_mode=active.lifecycle_mode,
            build_plan_json=plan.canonical_json,
            build_lease_sha256=plan.sha256,
            source_manifest_sha256=active.source_manifest_sha256,
            activation_sequence=active.activation_sequence,
            activation_sha256=active.activation_sha256,
            lineage_anchor_revision=anchor.catalog_revision,
            lineage_anchor_build_token=anchor.build_token,
            lineage_anchor_activation_sequence=anchor.activation_sequence,
            lineage_anchor_activation_sha256=anchor.activation_sha256,
            active_revisions_since_anchor=anchor.active_revisions_since,
            hot_producer_stream_id=hot.producer_stream_id,
            emitted_at=emitted_at,
        )

    @property
    def tenant_key(self) -> tuple[str, str]:
        return self.organization_id, self.workspace_id

    @property
    def document(self) -> Mapping[str, Any]:
        return {
            "organization_id": self.organization_id,
            "workspace_id": self.workspace_id,
            "catalog_epoch": self.catalog_epoch,
            "catalog_revision": self.catalog_revision,
            "build_token": self.build_token,
            "projection_version": self.projection_version,
            "lifecycle_mode": self.lifecycle_mode.value,
            "build_plan_json": self.build_plan_json,
            "build_lease_sha256": self.build_lease_sha256,
            "source_manifest_sha256": self.source_manifest_sha256,
            "activation_sequence": self.activation_sequence,
            "activation_sha256": self.activation_sha256,
            "lineage_anchor_revision": self.lineage_anchor_revision,
            "lineage_anchor_build_token": self.lineage_anchor_build_token,
            "lineage_anchor_activation_sequence": (
                self.lineage_anchor_activation_sequence
            ),
            "lineage_anchor_activation_sha256": (self.lineage_anchor_activation_sha256),
            "active_revisions_since_anchor": self.active_revisions_since_anchor,
            "hot_producer_stream_id": self.hot_producer_stream_id,
            "emitted_at": _time_text(self.emitted_at),
            "retirement_sha256": self.retirement_sha256,
        }


def producer_retirement_sha256(value: ProducerStateRetirement) -> str:
    return framed_sha256(
        PRODUCER_RETIREMENT_SHA_DOMAIN,
        value.organization_id,
        value.workspace_id,
        value.catalog_epoch,
        value.catalog_revision,
        value.build_token,
        value.projection_version,
        value.lifecycle_mode.value,
        value.build_plan_json,
        value.build_lease_sha256,
        value.source_manifest_sha256,
        value.activation_sequence,
        value.activation_sha256,
        value.lineage_anchor_revision,
        value.lineage_anchor_build_token,
        value.lineage_anchor_activation_sequence,
        value.lineage_anchor_activation_sha256,
        value.active_revisions_since_anchor,
        value.hot_producer_stream_id,
        _time_text(value.emitted_at),
    )


def encode_producer_retirements(
    values: Sequence[ProducerStateRetirement],
) -> bytes:
    ordered = tuple(sorted(values, key=lambda value: value.tenant_key))
    if not 1 <= len(ordered) <= MAX_PRODUCER_RETIREMENTS or len(
        {value.tenant_key for value in ordered}
    ) != len(ordered):
        raise ProducerRetirementError(
            "retirement document requires unique bounded tenant records"
        )
    document = {
        "format": PRODUCER_RETIREMENT_FORMAT,
        "version": PRODUCER_RETIREMENT_VERSION,
        "retirements": [value.document for value in ordered],
    }
    raw = (
        json.dumps(document, ensure_ascii=False, separators=(",", ":")).encode("utf-8")
        + b"\n"
    )
    if len(raw) > MAX_PRODUCER_RETIREMENT_BYTES:
        raise ProducerRetirementError("retirement document exceeds its byte limit")
    return raw


def decode_producer_retirements(raw: bytes) -> tuple[ProducerStateRetirement, ...]:
    if (
        not isinstance(raw, bytes)
        or len(raw) < 2
        or len(raw) > MAX_PRODUCER_RETIREMENT_BYTES
        or not raw.endswith(b"\n")
        or b"\n" in raw[:-1]
    ):
        raise ProducerRetirementError(
            "retirement document is not one bounded canonical JSON line"
        )
    try:
        decoded = json.loads(raw[:-1])
    except (UnicodeDecodeError, json.JSONDecodeError) as exc:
        raise ProducerRetirementError("retirement document is invalid JSON") from exc
    if not isinstance(decoded, dict) or tuple(decoded) != (
        "format",
        "version",
        "retirements",
    ):
        raise ProducerRetirementError("retirement document fields are invalid")
    records = decoded["retirements"]
    if (
        decoded["format"] != PRODUCER_RETIREMENT_FORMAT
        or decoded["version"] != PRODUCER_RETIREMENT_VERSION
        or not isinstance(records, list)
        or not 1 <= len(records) <= MAX_PRODUCER_RETIREMENTS
    ):
        raise ProducerRetirementError("retirement document format/count is invalid")
    values: list[ProducerStateRetirement] = []
    for index, raw_record in enumerate(records):
        if not isinstance(raw_record, dict) or tuple(raw_record) != _RECORD_FIELDS:
            raise ProducerRetirementError(
                f"retirement record {index} fields/order are invalid"
            )
        try:
            values.append(
                ProducerStateRetirement(
                    organization_id=raw_record["organization_id"],
                    workspace_id=raw_record["workspace_id"],
                    catalog_epoch=raw_record["catalog_epoch"],
                    catalog_revision=raw_record["catalog_revision"],
                    build_token=raw_record["build_token"],
                    projection_version=raw_record["projection_version"],
                    lifecycle_mode=CatalogLifecycleMode(raw_record["lifecycle_mode"]),
                    build_plan_json=raw_record["build_plan_json"],
                    build_lease_sha256=raw_record["build_lease_sha256"],
                    source_manifest_sha256=raw_record["source_manifest_sha256"],
                    activation_sequence=raw_record["activation_sequence"],
                    activation_sha256=raw_record["activation_sha256"],
                    lineage_anchor_revision=raw_record["lineage_anchor_revision"],
                    lineage_anchor_build_token=raw_record["lineage_anchor_build_token"],
                    lineage_anchor_activation_sequence=raw_record[
                        "lineage_anchor_activation_sequence"
                    ],
                    lineage_anchor_activation_sha256=raw_record[
                        "lineage_anchor_activation_sha256"
                    ],
                    active_revisions_since_anchor=raw_record[
                        "active_revisions_since_anchor"
                    ],
                    hot_producer_stream_id=raw_record["hot_producer_stream_id"],
                    emitted_at=_parse_time(raw_record["emitted_at"]),
                    retirement_sha256=raw_record["retirement_sha256"],
                )
            )
        except (TypeError, ValueError) as exc:
            raise ProducerRetirementError(
                f"retirement record {index} is invalid: {exc}"
            ) from exc
    result = tuple(values)
    if tuple(value.tenant_key for value in result) != tuple(
        sorted({value.tenant_key for value in result})
    ):
        raise ProducerRetirementError(
            "retirement records must be unique and tenant-sorted"
        )
    if encode_producer_retirements(result) != raw:
        raise ProducerRetirementError("retirement document is not canonical JSON")
    return result


class AtomicProducerStateRetirementFile:
    """Publish one monotonic active high-water per tenant on the shared volume."""

    def __init__(self, path: str | Path) -> None:
        self._path = Path(path)
        if not self._path.is_absolute():
            raise ValueError("producer retirement path must be absolute")
        try:
            parent = os.lstat(self._path.parent)
        except OSError as exc:
            raise ValueError("producer retirement parent is unavailable") from exc
        if not stat.S_ISDIR(parent.st_mode) or stat.S_ISLNK(parent.st_mode):
            raise ValueError("producer retirement parent must be a physical directory")

    @property
    def path(self) -> Path:
        return self._path

    def publish(self, value: ProducerStateRetirement) -> None:
        if not isinstance(value, ProducerStateRetirement):
            raise TypeError("value must be ProducerStateRetirement")
        current = {record.tenant_key: record for record in self._read()}
        previous = current.get(value.tenant_key)
        if previous is not None:
            previous_order = (
                previous.catalog_epoch,
                previous.activation_sequence,
                previous.catalog_revision,
            )
            value_order = (
                value.catalog_epoch,
                value.activation_sequence,
                value.catalog_revision,
            )
            if value_order < previous_order:
                raise ProducerRetirementError("retirement high-water regressed")
            if value_order == previous_order:
                if _activation_identity(value) != _activation_identity(previous):
                    raise ProducerRetirementError(
                        "retirement conflicts at one activation high-water"
                    )
                return
            if value.catalog_epoch == previous.catalog_epoch and (
                value.activation_sequence <= previous.activation_sequence
                or value.catalog_revision <= previous.catalog_revision
            ):
                raise ProducerRetirementError(
                    "retirement revision/activation sequence is not monotonic"
                )
        current[value.tenant_key] = value
        self._write(tuple(current.values()))

    def _read(self) -> tuple[ProducerStateRetirement, ...]:
        try:
            info = os.lstat(self._path)
        except FileNotFoundError:
            return ()
        except OSError as exc:
            raise ProducerRetirementError(
                "cannot inspect producer retirement file"
            ) from exc
        if (
            not stat.S_ISREG(info.st_mode)
            or stat.S_ISLNK(info.st_mode)
            or info.st_mode & 0o022
            or not 2 <= info.st_size <= MAX_PRODUCER_RETIREMENT_BYTES
        ):
            raise ProducerRetirementError(
                "producer retirement file type/mode/size is unsafe"
            )
        try:
            raw = self._path.read_bytes()
        except OSError as exc:
            raise ProducerRetirementError(
                "cannot read producer retirement file"
            ) from exc
        return decode_producer_retirements(raw)

    def _write(self, values: Sequence[ProducerStateRetirement]) -> None:
        raw = encode_producer_retirements(values)
        descriptor, temporary_name = tempfile.mkstemp(
            prefix=".producer-retirement-",
            dir=self._path.parent,
        )
        temporary = Path(temporary_name)
        keep = True
        try:
            os.fchmod(descriptor, 0o600)
            written = os.write(descriptor, raw)
            if written != len(raw):
                raise ProducerRetirementError(
                    "producer retirement file was only partially written"
                )
            os.fsync(descriptor)
            os.close(descriptor)
            descriptor = -1
            os.replace(temporary, self._path)
            keep = False
            directory_fd = os.open(self._path.parent, os.O_RDONLY | os.O_DIRECTORY)
            try:
                os.fsync(directory_fd)
            finally:
                os.close(directory_fd)
        finally:
            if descriptor >= 0:
                os.close(descriptor)
            if keep:
                try:
                    temporary.unlink()
                except FileNotFoundError:
                    pass


def _activation_identity(value: ProducerStateRetirement) -> ProducerStateRetirement:
    return replace(
        value,
        emitted_at=datetime(1970, 1, 1, tzinfo=UTC),
        retirement_sha256=_ZERO_SHA256,
    )


__all__ = [
    "AtomicProducerStateRetirementFile",
    "MAX_PRODUCER_RETIREMENT_BYTES",
    "MAX_PRODUCER_RETIREMENTS",
    "PRODUCER_RETIREMENT_FILE_NAME",
    "PRODUCER_RETIREMENT_FORMAT",
    "PRODUCER_RETIREMENT_SHA_DOMAIN",
    "PRODUCER_RETIREMENT_VERSION",
    "ProducerRetirementError",
    "ProducerStateRetirement",
    "decode_producer_retirements",
    "encode_producer_retirements",
    "producer_retirement_sha256",
]

File: fi-collector/pkg/propertycatalog/producer_retirement.go
package propertycatalog

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"
)

const (
	producerRetirementFormat    = "futureagi.property-catalog-producer-state-retirements"
	producerRetirementVersion   = uint16(1)
	producerRetirementFileName  = "producer-state-retirements-v1.json"
	producerRetirementSHADomain = "futureagi.property-catalog.producer-state-retirement.v1"
	maxProducerRetirementBytes  = 64 << 20
)

type producerRetirementDocument struct {
	Format      string                    `json:"format"`
	Version     uint16                    `json:"version"`
	Retirements []ProducerStateRetirement `json:"retirements"`
}

// ProducerStateRetirement is the Python control plane's activation-proven
// high-water. It is intentionally not derived from the mutable fence file.
// The canonical active build plan and activation lineage are reread from
// ClickHouse before Python atomically publishes this record.
type ProducerStateRetirement struct {
	OrganizationID                  string `json:"organization_id"`
	WorkspaceID                     string `json:"workspace_id"`
	CatalogEpoch                    uint16 `json:"catalog_epoch"`
	CatalogRevision                 uint64 `json:"catalog_revision"`
	BuildToken                      string `json:"build_token"`
	ProjectionVersion               uint16 `json:"projection_version"`
	LifecycleMode                   string `json:"lifecycle_mode"`
	BuildPlanJSON                   string `json:"build_plan_json"`
	BuildLeaseSHA256                string `json:"build_lease_sha256"`
	SourceManifestSHA256            string `json:"source_manifest_sha256"`
	ActivationSequence              uint64 `json:"activation_sequence"`
	ActivationSHA256                string `json:"activation_sha256"`
	LineageAnchorRevision           uint64 `json:"lineage_anchor_revision"`
	LineageAnchorBuildToken         string `json:"lineage_anchor_build_token"`
	LineageAnchorActivationSequence uint64 `json:"lineage_anchor_activation_sequence"`
	LineageAnchorActivationSHA256   string `json:"lineage_anchor_activation_sha256"`
	ActiveRevisionsSinceAnchor      uint64 `json:"active_revisions_since_anchor"`
	HotProducerStreamID             string `json:"hot_producer_stream_id"`
	EmittedAt                       string `json:"emitted_at"`
	RetirementSHA256                string `json:"retirement_sha256"`
}

type producerRetirementTenant struct {
	organizationID string
	workspaceID    string
}

func producerRetirementSHA256(value ProducerStateRetirement) string {
	return framedSHA256(
		producerRetirementSHADomain,
		value.OrganizationID,
		value.WorkspaceID,
		uint64(value.CatalogEpoch),
		value.CatalogRevision,
		value.BuildToken,
		uint64(value.ProjectionVersion),
		value.LifecycleMode,
		value.BuildPlanJSON,
		value.BuildLeaseSHA256,
		value.SourceManifestSHA256,
		value.ActivationSequence,
		value.ActivationSHA256,
		value.LineageAnchorRevision,
		value.LineageAnchorBuildToken,
		value.LineageAnchorActivationSequence,
		value.LineageAnchorActivationSHA256,
		value.ActiveRevisionsSinceAnchor,
		value.HotProducerStreamID,
		value.EmittedAt,
	)
}

func loadProducerRetirements(directory string) (map[producerRetirementTenant]ProducerStateRetirement, error) {
	path := filepath.Join(directory, producerRetirementFileName)
	info, err := os.Lstat(path)
	if errors.Is(err, os.ErrNotExist) {
		return map[producerRetirementTenant]ProducerStateRetirement{}, nil
	}
	if err != nil {
		return nil, fmt.Errorf("propertycatalog: inspect producer retirement proof: %w", err)
	}
	if !info.Mode().IsRegular() || info.Mode()&os.ModeSymlink != 0 || info.Mode().Perm() != 0o600 ||
		info.Size() < 2 || info.Size() > maxProducerRetirementBytes {
		return nil, errors.New("propertycatalog: producer retirement proof type, mode, or size is unsafe")
	}
	raw, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("propertycatalog: read producer retirement proof: %w", err)
	}
	if raw[len(raw)-1] != '\n' || bytes.Contains(raw[:len(raw)-1], []byte{'\n'}) {
		return nil, errors.New("propertycatalog: producer retirement proof is not one canonical JSON line")
	}
	body := raw[:len(raw)-1]
	decoder := json.NewDecoder(bytes.NewReader(body))
	decoder.DisallowUnknownFields()
	var document producerRetirementDocument
	if err := decoder.Decode(&document); err != nil {
		return nil, fmt.Errorf("propertycatalog: decode producer retirement proof: %w", err)
	}
	if err := requireJSONEOF(decoder); err != nil {
		return nil, err
	}
	canonical, err := json.Marshal(document)
	if err != nil || !bytes.Equal(canonical, body) {
		return nil, errors.New("propertycatalog: producer retirement proof is not canonical JSON")
	}
	if document.Format != producerRetirementFormat || document.Version != producerRetirementVersion ||
		len(document.Retirements) == 0 {
		return nil, errors.New("propertycatalog: producer retirement proof format/version/count is invalid")
	}
	if !sort.SliceIsSorted(document.Retirements, func(i, j int) bool {
		left, right := document.Retirements[i], document.Retirements[j]
		if left.OrganizationID != right.OrganizationID {
			return left.OrganizationID < right.OrganizationID
		}
		return left.WorkspaceID < right.WorkspaceID
	}) {
		return nil, errors.New("propertycatalog: producer retirements are not tenant-sorted")
	}
	result := make(map[producerRetirementTenant]ProducerStateRetirement, len(document.Retirements))
	for index, retirement := range document.Retirements {
		if err := validateProducerRetirement(retirement); err != nil {
			return nil, fmt.Errorf("propertycatalog: producer retirement %d: %w", index, err)
		}
		key := producerRetirementTenant{retirement.OrganizationID, retirement.WorkspaceID}
		if _, duplicate := result[key]; duplicate {
			return nil, errors.New("propertycatalog: producer retirement proof contains a duplicate tenant")
		}
		result[key] = retirement
	}
	return result, nil
}

func validateProducerRetirement(value ProducerStateRetirement) error {
	for label, candidate := range map[string]string{
		"retirement organization":        value.OrganizationID,
		"retirement workspace":           value.WorkspaceID,
		"retirement build token":         value.BuildToken,
		"retirement anchor build token":  value.LineageAnchorBuildToken,
		"retirement hot producer stream": value.HotProducerStreamID,
	} {
		if err := validateCanonicalUUID(label, candidate); err != nil {
			return err
		}
	}
	if value.CatalogEpoch == 0 || value.CatalogRevision == 0 || value.ProjectionVersion == 0 ||
		value.ActivationSequence == 0 || value.LineageAnchorRevision == 0 ||
		value.LineageAnchorActivationSequence == 0 || value.ActiveRevisionsSinceAnchor > 2048 {
		return errors.New("retirement revision, activation, or lineage bound is invalid")
	}
	for label, digest := range map[string]string{
		"build lease":       value.BuildLeaseSHA256,
		"source manifest":   value.SourceManifestSHA256,
		"activation":        value.ActivationSHA256,
		"anchor activation": value.LineageAnchorActivationSHA256,
		"retirement":        value.RetirementSHA256,
	} {
		if !isLowerSHA256(digest) {
			return fmt.Errorf("%s digest is invalid", label)
		}
	}
	emitted, err := time.Parse(dateTime64Layout, value.EmittedAt)
	if err != nil || emitted.UTC().Format(dateTime64Layout) != value.EmittedAt {
		return errors.New("retirement emission time is not canonical UTC")
	}
	if value.ActivationSequence < value.LineageAnchorActivationSequence ||
		value.ActivationSequence-value.LineageAnchorActivationSequence != value.ActiveRevisionsSinceAnchor {
		return errors.New("retirement activation depth differs from its lineage anchor")
	}
	snapshot := value.LifecycleMode == "initial_backfill" || value.LifecycleMode == "full_repair"
	switch {
	case snapshot:
		if value.LineageAnchorRevision != value.CatalogRevision ||
			value.LineageAnchorBuildToken != value.BuildToken ||
			value.LineageAnchorActivationSequence != value.ActivationSequence ||
			value.LineageAnchorActivationSHA256 != value.ActivationSHA256 ||
			value.ActiveRevisionsSinceAnchor != 0 {
			return errors.New("snapshot retirement is not its own lineage anchor")
		}
	case value.LifecycleMode == "incremental":
		if value.LineageAnchorRevision >= value.CatalogRevision ||
			value.LineageAnchorActivationSequence >= value.ActivationSequence {
			return errors.New("incremental retirement has no earlier lineage anchor")
		}
	default:
		return errors.New("retirement lifecycle mode is unsupported")
	}
	request := DeliveryLeaseRequest{
		OrganizationID: value.OrganizationID, WorkspaceID: value.WorkspaceID,
		CatalogEpoch: value.CatalogEpoch, CatalogRevision: value.CatalogRevision,
		BuildToken: value.BuildToken, ProjectionVersion: value.ProjectionVersion,
		SourceAdapter: AdapterSpanAttribute, ProducerStreamID: value.HotProducerStreamID,
		EnvelopeVersion: EnvelopeVersion, Sequence: 1,
	}
	evidence, err := validateBuildPlan(value.BuildPlanJSON, value.BuildLeaseSHA256, request)
	if err != nil {
		return fmt.Errorf("retirement build plan: %w", err)
	}
	if evidence.StreamRole != "hot_values" {
		return errors.New("retirement stream is not the active hot-values stream")
	}
	var plan buildPlanDocumentJSON
	if err := json.Unmarshal([]byte(value.BuildPlanJSON), &plan); err != nil {
		return err
	}
	prefix := value.LifecycleMode + "_"
	for _, stream := range plan.Streams {
		if !strings.HasPrefix(stream.SourceCutoff.Label, prefix) {
			return errors.New("retirement lifecycle mode differs from its build plan")
		}
	}
	if value.RetirementSHA256 != producerRetirementSHA256(value) {
		return errors.New("retirement digest does not match its fields")
	}
	return nil
}

func (r *HotRuntime) compactProducerState(ctx context.Context, fences []RevisionFence) error {
	if r == nil || r.publisher == nil || r.publisher.state == nil || ctx == nil {
		return errors.New("propertycatalog: producer retirement requires a runtime context")
	}
	if err := ctx.Err(); err != nil {
		return err
	}
	r.retirementMu.Lock()
	defer r.retirementMu.Unlock()
	retirements, err := loadProducerRetirements(r.cfg.SpoolDirectory)
	if err != nil || len(retirements) == 0 {
		return err
	}
	current := make(map[producerRetirementTenant]RevisionFence, len(fences))
	for _, fence := range fences {
		key := producerRetirementTenant{fence.OrganizationID, fence.WorkspaceID}
		if _, duplicate := current[key]; duplicate {
			return errors.New("propertycatalog: current fence contains duplicate retirement tenant")
		}
		current[key] = fence
	}
	pendingEnvelopes, err := r.spool.PendingEnvelopes()
	if err != nil {
		return err
	}
	pending := make(map[streamKey]struct{}, len(pendingEnvelopes))
	for _, envelope := range pendingEnvelopes {
		snapshot := envelope.Snapshot()
		pending[streamKey{
			organizationID: snapshot.OrganizationID, workspaceID: snapshot.WorkspaceID,
			epoch: snapshot.CatalogEpoch, revision: snapshot.CatalogRevision,
			buildToken: snapshot.BuildToken, adapter: snapshot.SourceAdapter,
			streamID: snapshot.ProducerStreamID,
		}] = struct{}{}
	}
	covered := func(key streamKey, projection uint16) (bool, error) {
		retirement, proven := retirements[producerRetirementTenant{
			key.organizationID, key.workspaceID,
		}]
		fence, assigned := current[producerRetirementTenant{
			key.organizationID, key.workspaceID,
		}]
		if !proven || !assigned || key.epoch != retirement.CatalogEpoch ||
			key.revision > retirement.CatalogRevision || projection != retirement.ProjectionVersion ||
			key.adapter != AdapterSpanAttribute || key.streamID != retirement.HotProducerStreamID ||
			fence.CatalogEpoch != key.epoch || fence.CatalogRevision <= key.revision ||
			fence.ProjectionVersion != projection {
			return false, nil
		}
		if key.revision == retirement.CatalogRevision && key.buildToken != retirement.BuildToken {
			return false, errors.New("propertycatalog: active retirement conflicts with producer checkpoint build")
		}
		return true, nil
	}
	remove := make(map[streamKey]struct{})
	checkpoints := r.publisher.state.snapshot()
	for key, checkpoint := range checkpoints {
		isCovered, coverErr := covered(key, checkpoint.ProjectionVersion)
		if coverErr != nil {
			return coverErr
		}
		if !isCovered || !checkpoint.Terminal {
			continue
		}
		if _, exists := pending[key]; exists {
			continue
		}
		remove[key] = struct{}{}
	}
	r.mu.Lock()
	defer r.mu.Unlock()
	for key := range remove {
		if r.pendingAdmissions[key] != 0 {
			return errors.New("propertycatalog: producer retirement crossed a pending admission")
		}
	}
	cleanupDrains := make(map[streamKey]struct{})
	for key := range r.drains {
		if _, hasCheckpoint := checkpoints[key]; hasCheckpoint {
			if _, willRemove := remove[key]; !willRemove {
				continue
			}
		}
		if _, hasPending := pending[key]; hasPending || r.pendingAdmissions[key] != 0 {
			continue
		}
		isCovered, coverErr := covered(key, r.cfg.ProjectionVersion)
		if coverErr != nil {
			return coverErr
		}
		if isCovered {
			cleanupDrains[key] = struct{}{}
		}
	}
	if len(remove) != 0 {
		if err := r.publisher.state.remove(remove); err != nil {
			return err
		}
	}
	for key := range remove {
		delete(r.tails, key)
		delete(r.pendingAdmissions, key)
		delete(r.admissionPoison, key)
		delete(r.drains, key)
	}
	// A prior successful compaction may have persisted the empty checkpoint
	// document immediately before a crash. loadDrainSafety then restores the old
	// ready marker from the separately durable proof. Clear only such checkpoint-
	// free, spool-free markers covered by the same activation high-water; no
	// producer-state write is needed because the checkpoint is already absent.
	for key := range cleanupDrains {
		delete(r.drains, key)
		delete(r.admissionPoison, key)
	}
	return nil
}
